craft_parts/executor/part_handler.py

```diff
@@ -380,6 +383,92 @@ def _update_build(self, step_info: StepInfo) -> None:
 
         self._organize(overwrite=True)
 
+    def _migrate_overlay_files_to_stage(self) -> None:
+        """Stage overlay files create state.
+
+        Files and directories are migrated from overlay to stage based on a
+        list of visible overlay entries, converting overlayfs whiteout files
+        and opaque dirs to OCI.
+        """
+        stage_overlay_state_path = states.get_overlay_migration_state_path(
+            self._part.overlay_dir, Step.STAGE
+        )
+        if stage_overlay_state_path.exists():
+            logger.debug("stage overlay migration state exists")
+            return
+
+        if self._part not in parts_with_overlay(part_list=self._part_list):
+            return
+
+        logger.debug("staging overlay files")
+        migrated_files: Set[str] = set()
+        migrated_dirs: Set[str] = set()
+
+        # process layers from top to bottom
+        for part in reversed(self._part_list):
+            if not part.has_overlay:
+                continue
+
+            logger.debug("migrate part %r layer to stage", part.name)
+            visible_files, visible_dirs = overlays.visible_in_layer(
+                part.part_layer_dir, part.stage_dir
+            )
+            layer_files, layer_dirs = migration.migrate_files(
+                files=visible_files,
+                dirs=visible_dirs,
+                srcdir=part.part_layer_dir,
+                destdir=part.stage_dir,
+                oci_translation=True,
+            )
+            migrated_files |= layer_files
+            migrated_dirs |= layer_dirs
+
+        state = MigrationState(files=migrated_files, directories=migrated_dirs)
+        state.write(stage_overlay_state_path)
+
+    def _migrate_overlay_files_to_prime(self) -> None:
+        """Prime overlay files and create state.
+
+        Files and directories are migrated from stage to prime based on a list
+        of visible overlay entries, including OCI-compatible whiteout files and
+        opaque directories.
+        """
+        prime_overlay_state_path = states.get_overlay_migration_state_path(
+            self._part.overlay_dir, Step.PRIME
+        )
+        if prime_overlay_state_path.exists():
```

Comment:
Overlay data is migrated to stage and prime when the first part declaring overlay parameters is migrated. This operation creates the overlay migration state. To prevent additional, unnecessary migrations, parts that contain overlay parameters verify if the migration state already exists before attempting to migrate overlay data.

Reply:
Can you document this in code?

Reply:
Added comment explaining when overlay data is migrated.